### PR TITLE
DDF-6360 Upgrade to CXF 3.4.0

### PR DESCRIPTION
**Pull request by SmithJosh**
#### What does this PR do?
Upgrades to CXF 3.4.0, and along with it upgrades a bunch of other dependencies, including:
* OpenSAML -> 3.4.5
* WSS4J -> 2.3.0
* Santuario (xmlsec) -> 2.2.0
* JAXB API -> Jakarta JAXB API
* Sun JAXB impl to Glassfish JAXB impl
* Javax Annotation API -> Jakarta Annotation API
* JavaBeans Activation Framework API -> Jakarta Activation API

#### Who is reviewing it? 

@bakejeyner @blen-desta @emmberk @garrettfreibott @stustison

#### Select relevant component teams: 
@codice/security 

#### How should this be tested?

The major issues I ran into when upgrading were very slow installation times and JAXB errors. With that in mind, verify that:. With that in mind, verify that:

* Installation is not abnormally slow
* Anything using JAXB still works (XML marshalling, unmarshalling)
  * Adding and querying sources: CSW & OpenSearch (loopback is fine) and WFS (I used http://geomap.reteunitaria.piemonte.it/ws/gsareprot/rp-01/areeprotwfs/wfs_gsareprot_1, which is WFS 1.1.0 and just tested that the source is shown as active)
  * XML transformer: ingest an XML document with geometry, then download it via the xml transformer

#### Any background context you want to provide?
This is part of an ongoing effort to reduce the CVE level in DDF

#### What are the relevant tickets?
Fixes: #6360

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Threat Dragon models
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: catalog/spatial/wfs/1.1.0/spatial-wfs-v1_1_0-converter/pom.xml [hunk 1]
```diff
@@ -101,6 +94,7 @@
             <artifactId>spatial-wfs-v1_1_0-common</artifactId>
             <version>${project.version}</version>
             <exclusions>
+                <!-- FIXME: There's no cxf-api artifact anymore -->
```
Comment (emmberk) [inline catalog/spatial/wfs/1.1.0/spatial-wfs-v1_1_0-converter/pom.xml:97]: ✏️ Do these FIXMEs need to be addressed?

File: distribution/ddf-common/src/main/resources/etc/jre.properties [hunk 2]
```diff
@@ -422,12 +422,12 @@ jre-1.8= \
...
- javax.xml.bind.annotation.adapters;version="2.2.8", \
- javax.xml.bind.attachment;version="2.2.8", \
- javax.xml.bind.helpers;version="2.2.8", \
- javax.xml.bind.util;version="2.2.8", \
+# javax.xml.bind;version="2.3.0", \
+# javax.xml.bind.annotation;version="2.3.0", \
+# javax.xml.bind.annotation.adapters;version="2.3.0", \
+# javax.xml.bind.attachment;version="2.3.0", \
+# javax.xml.bind.helpers;version="2.3.0", \
+# javax.xml.bind.util;version="2.3.0", \
```
Comment (emmberk) [inline distribution/ddf-common/src/main/resources/etc/jre.properties:430]: ✏️ Why are these commented out?

File: platform/security/pep/security-pep-interceptor/src/main/java/ddf/security/pep/interceptor/SecurityAssertionStore.java [hunk 3]
```diff
@@ -116,12 +117,16 @@ public static SecurityAssertion getSecurityAssertion(Message message) {
...
 
   /**
-   * Return the TokenStore associated with this message.
+   * Return the TokenStore associated with this message
    *
    * @param message
-   * @return TokenStore
+   * @return the TokenStore, or null if it could not be retrieved
    */
   public static TokenStore getTokenStore(Message message) {
```
Comment (emmberk) [inline platform/security/pep/security-pep-interceptor/src/main/java/ddf/security/pep/interceptor/SecurityAssertionStore.java:125]: ✏️ Add `@Nullable` annotation

**Review by emmberk — APPROVED** [review]

File: libs/bundles/libs-bundles-opensaml/pom.xml [hunk 4]
```diff
@@ -1,223 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (SmithJosh) [inline libs/bundles/libs-bundles-opensaml/pom.xml:1]: After previous discussions about removing this, I ended up going ahead and doing it. The reason was that the Servicemix OpenSAML bundle now has no version constraint on the Guava version is uses. See: https://github.com/apache/servicemix-bundles/blob/master/opensaml-3.4.5/pom.xml#L63

This means it will pick up the Guava version that DDF uses automatically; we don't need to force the version here. Note, though, that we have to make sure that there is only one Guava version available or there will be uses constraint violations.

I'm open to keeping this bundle if we think that's a better option. But these changes work just fine without it.

File: platform/security/core/security-core-impl/pom.xml [hunk 5]
```diff
@@ -252,8 +232,6 @@
                             !org.mockito*,
                             !org.hamcrest*,
                             !org.junit*,
-                            org.opensaml.xmlsec.signature.support.provider,
-                            org.opensaml.*;version="${opensaml.version}",
```
Comment (SmithJosh) [inline platform/security/core/security-core-impl/pom.xml:256]: bnd figures out these imports just fine

File: platform/security/pep/security-pep-interceptor/src/main/java/ddf/security/pep/interceptor/SecurityAssertionStore.java [hunk 6]
```diff
@@ -116,12 +117,16 @@ public static SecurityAssertion getSecurityAssertion(Message message) {
...
-    return TokenStoreUtils.getTokenStore(message);
+    try {
+      return TokenStoreUtils.getTokenStore(message);
+    } catch (TokenStoreException e) {
+      return null;
+    }
```
Comment (SmithJosh) [inline platform/security/pep/security-pep-interceptor/src/main/java/ddf/security/pep/interceptor/SecurityAssertionStore.java:130]: The only code change for this PR 🎉 

If this is null, we just return a new SecurityAssertionSaml instance. I'm not sure if that's what we actually want, as the code looks like we were not expecting the token store to be null. Would appreciate another set of eyes.

https://github.com/codice/ddf/blob/master/platform/security/pep/security-pep-interceptor/src/main/java/ddf/security/pep/interceptor/SecurityAssertionStore.java#L115

File: catalog/spatial/wfs/1.1.0/spatial-wfs-v1_1_0-converter/pom.xml (hunk 1 above)
Comment (SmithJosh) [reply catalog/spatial/wfs/1.1.0/spatial-wfs-v1_1_0-converter/pom.xml:97]: I removed the exclusions. Since the artifact doesn't exist the exclusions didn't do anything and it still worked

File: catalog/transformer/catalog-transformer-xlsx/pom.xml [hunk 7]
```diff
@@ -84,8 +75,7 @@
                     <instructions>
                         <Embed-Dependency>
                             catalog-core-api-impl,
-                            platform-util,
-                            security-rest-clientapi
+                            platform-util
```
Comment (stustison) [inline catalog/transformer/catalog-transformer-xlsx/pom.xml:78]: platform-util is a bundle so it shouldn't be embedded. I must have missed this when I removed it from all the poms or maybe it was added after I changed things.

File: libs/bundles/libs-bundles-opensaml/pom.xml (hunk 4 above)
Comment (stustison) [reply libs/bundles/libs-bundles-opensaml/pom.xml:1]: I vote we remove it.

File: platform/security/pep/security-pep-interceptor/src/main/java/ddf/security/pep/interceptor/SecurityAssertionStore.java (hunk 6 above)
Comment (stustison) [reply platform/security/pep/security-pep-interceptor/src/main/java/ddf/security/pep/interceptor/SecurityAssertionStore.java:130]: It will work correctly the way it is and the PEP still fails closed as long as the token for the assertion is null, which it will be if you just create an empty SAML assertion. However, it might be more correct if we just return an exception if the token store is null or in the case where we might flow down to returning that empty SecurityAssertionSaml instance. I think all that empty instance does is trigger the error condition anyways, so we may as well just jump there. That way we can avoid someone changing the PEP in the future and this possibly failing open at some point.

File: catalog/transformer/catalog-transformer-xlsx/pom.xml (hunk 7 above)
Comment (SmithJosh) [reply catalog/transformer/catalog-transformer-xlsx/pom.xml:78]: ok I'll remove it

File: platform/security/pep/security-pep-interceptor/src/main/java/ddf/security/pep/interceptor/SecurityAssertionStore.java (hunk 6 above)
Comment (SmithJosh) [reply platform/security/pep/security-pep-interceptor/src/main/java/ddf/security/pep/interceptor/SecurityAssertionStore.java:130]: Ok thanks for confirming. I'm inclined to keep it as-is in that case since the PEP still denies access. I checked and the PEP has tests for null tokens and null assertions so that should raise some flags if the it starts failing open.

**Comment by SmithJosh** [comment]
build now

**Comment by SmithJosh** [comment]
Rebased. build now

**Comment by SmithJosh** [comment]
build now

File: libs/bundles/libs-bundles-opensaml/pom.xml (hunk 4 above)
Comment (Lambeaux) [reply libs/bundles/libs-bundles-opensaml/pom.xml:1]: (❗ ) If only one Guava version can be available, then this may break some Camel features. Check out the Camel features file: 
```
<feature name="camel-protobuf" version="${camel.version}" start-level="50">
        <feature version="${camel.version}">camel-core</feature>
        <bundle dependency="true">mvn:com.google.protobuf/protobuf-java/${protobuf-version}</bundle>
        <bundle dependency="true">mvn:com.google.protobuf/protobuf-java-util/${protobuf-version}</bundle>
        <bundle dependency="true">mvn:commons-io/commons-io/${commons-io-version}</bundle>
        <bundle dependency="true">mvn:com.google.guava/guava/${camel.guava.version}</bundle>
        <bundle dependency="true">mvn:com.google.code.gson/gson/${gson.version}</bundle>
        <bundle>mvn:org.apache.camel/camel-protobuf/${camel.version}</bundle>
    </feature>
```
It appears the Camel version of Guava is lower than DDF's version of Guava. 
```
<camel.guava.version>19.0</camel.guava.version>
```

**Review by bakejeyner — APPROVED** [review]

File: libs/bundles/libs-bundles-opensaml/pom.xml (hunk 4 above)
Comment (SmithJosh) [reply libs/bundles/libs-bundles-opensaml/pom.xml:1]: Turns out this is actually a lot more nuanced than I was thinking. Interestingly, installing the feature you linked also causes uses constraint violations on 2.26.0. It's not just these changes - *DDF cannot have two Gauva versions installed at the same time*. Here's a sample of the constraint violations you see if it does:

```
2020-10-05T08:56:24,035 | DEBUG | tures-3-thread-1 | ResolverImpl                     | ternal.resolver.Slf4jResolverLog   45 | 17 - org.apache.karaf.features.core - 4.2.9 | Candidate permutation failed due to a conflict between imports; will try another if possible. (Uses constraint violation. Unable to resolve resource solr-factory-impl [solr-factory-impl/2.26.0] because it is exposed to package 'com.google.common.collect' from resources com.google.guava [com.google.guava/25.1.0.jre] and com.google.guava [com.google.guava/25.1.0.jre] via two dependency chains.

Chain 1:
  solr-factory-impl [solr-factory-impl/2.26.0]
    import: (&(osgi.wiring.package=com.google.common.collect)(version>=25.1.0)(!(version>=26.0.0)))
     |
    export: osgi.wiring.package: com.google.common.collect
  com.google.guava [com.google.guava/25.1.0.jre]

Chain 2:
  solr-factory-impl [solr-factory-impl/2.26.0]
    import: (&(osgi.wiring.package=org.codice.ddf.platform.util)(version>=2.26.0)(!(version>=3.0.0)))
     |
    export: osgi.wiring.package=org.codice.ddf.platform.util; uses:=com.google.common.io
  platform-util [platform-util/2.26.0]
    import: (&(osgi.wiring.package=com.google.common.io)(version>=25.1.0)(!(version>=26.0.0)))
     |
    export: osgi.wiring.package: com.google.common.io; uses:=com.google.common.collect
    export: osgi.wiring.package=com.google.common.collect
  com.google.guava [com.google.guava/25.1.0.jre])
```

A couple notes:
* This isn't actually an error, just a Karaf debug message. I found that interesting - there are non-fatal uses constraint violations. However, while it doesn't break anything Karaf takes significantly longer to resolve bundle wiring. On my machine, a `profile:install standard` took about 6 minutes compared to 1 minute normally.
* The error message here is a bit confusing since it references 25.1.0.jre for both dependency chains. It's really an issue with the uses constrains being satisfied by both versions of Guava present.
* Part of the problem is DDFs own leaky abstractions: platform-util exposes Guava in its API. Meaning anything that uses Guava and platform-util will have uses constraints violations if two Guava versions are present.

The only reason DDF 2.26.0 doesn't have these issues by default is that it doesn't install multiple Guava versions. That surprised me because the CXF and DDF versions differ. I dug into it a bit and it's because CXF uses the dependency flag in its feature descriptor. From the Karaf mailing list, here's a description of the dependency flag:

> The role of the dependency attribute is to mark that a bundle is a dependency. If a dependency is already satisfied (an existing bundle already exports the same packages/version) then it doesn't get installed. This behavior happens if the declared feature resolver is installed (e.g. obr is installed).

The wrapped OpenSAML bundle imports Guava packages with the version matching the DDF Guava version. So when when Karaf installs OpenSAML, it sees that the imports are already satisfied and so doesn't try to install the CXF Guava version (note, CXF itself doesn't import Guava. It only pulls it in to satisfy OpenSAML). Now that we're using the OpenSAML bundle without a version constraint, it appears Karaf doesn't think the imports are already satisfied and so installs the new version.

To summarize: with these changes, the only difference from master is that the CXF and DDF Guava versions must match to prevent uses constraint violations. With or without these changes, the presence of multiple Guava versions is problematic and will cause non-fatal uses constraint violations and significantly longer installation times.

Comment (Lambeaux) [reply libs/bundles/libs-bundles-opensaml/pom.xml:1]: That's some quality background and investigation. Even though we can get the container to start I still believe that 6 minutes to start is just as bad as a borked container that doesn't work at all because you have to wait far too long (synchronously) to learn if you've succeeded or failed. So for future readers of this - know that wiring issues like the above are basically fatal even if the container is capable of recovering.

To add to the above ☝️ the OpenSAML bundle needed to be wrapped because, as of October of last year, no new versions existed without the transitive version constraint. As @SmithJosh pointed out, this has recently been resolved with several new artifacts released from servicemix. 

Additionally, between the time I upgraded Guava and this change, the class path complexity has increased. Platform-util is now a bundle which itself exposes Guava. 
Guava upgrade: https://github.com/codice/ddf/pull/5379
Platform util: https://github.com/codice/ddf/pull/5776

This means that, while the Camel feature may have worked fine last year, it no longer does. Likely because of a new class path collision with platform util's Guava and the code that Camel needs. 

Given the wrapper is no longer serving any purpose I'm +1 to remove it now, as well. 👍

**Review by Lambeaux — APPROVED** [review]

File: libs/bundles/libs-bundles-opensaml/pom.xml (hunk 4 above)
Comment (SmithJosh) [reply libs/bundles/libs-bundles-opensaml/pom.xml:1]: Thanks @Lambeaux. Completely agree with this:
> Even though we can get the container to start I still believe that 6 minutes to start is just as bad as a borked container that doesn't work at all because you have to wait far too long (synchronously) to learn if you've succeeded or failed

I created https://github.com/codice/ddf/issues/6365 to fix it

**Comment by SmithJosh** [comment]
Rebased. build now

**Comment by SmithJosh** [comment]
build now

**Comment by SmithJosh** [comment]
build now

**Comment by SmithJosh** [comment]
Unrelated error:
```
[2020-10-06T14:36:51.939Z] [ERROR] DOCKER> Unable to build image [docker.io/codice/ddf:2.27.0-SNAPSHOT] : "The command '/bin/sh -c apk add --no-cache unzip' returned a non-zero code: 1"  ["The command '/bin/sh -c apk add --no-cache unzip' returned a non-zero code: 1" ]
```
build now

**Comment by SmithJosh** [comment]
Itest failure. I suspect a flakey test:
```
[2020-10-06T16:27:00.668Z] [ERROR] Failures: 
[2020-10-06T16:27:00.669Z] [ERROR]   TestFederation.testSingleUserDownloadSameProductSyncAndAsync:1884->verifyCswStubCall:1961 Expected to happen 1 time(s), but happened 2 times instead
```
build now

**Comment by SmithJosh** [comment]
Another seemingly unrelated failure. Tried running itests on my machine and they passed 😕 
```
[2020-10-06T18:03:02.187Z] [ERROR] ddf.test.itests.catalog.TestFederation.testFederatedNegativeSpatial  Time elapsed: 300.07 s  <<< ERROR!
11:03:02  org.ops4j.pax.swissbox.tracker.ServiceLookupException: gave up waiting for service org.ops4j.pax.exam.ProbeInvoker
```
build now

**Review by stustison — APPROVED** [review]

**Comment by jlcsmith** [comment]
🎉 hero successful 🎉

verified normal build/install
verified ingest of XML
verified CSW/WFS sources
